### ISSUE TITLE
Printed Output Has Significant Flaws

### ISSUE THREAD

**Issue by ssokolow** [ralfbiedert/cheats.rs#117]
I've noticed that the site redesign has caused serious breakages in both the print and cached PDF output.

* The `noprint` class on the "Fork Me on GitHub ribbon" has no effect.
* Entire sections of the reference which I'd *want* to print are hidden inside `<noprint>` elements.
* Tab widgets which do print are not unrolled, denying access to all but their default tabs.
* Expandable explanations in the memory diagrams print collapsed.
* If I turn off `noprint` to access the stuff I want, bits of your diagrams show up in the top-left of page 1 in print mode. (Honestly, I'd suggest using inline SVG for those rather than abusing things like `position: absolute` and `<table>` to do illustration.)
* If I uncollapse the tab widgets the most obvious way, they overdraw the next section down.

This is how far I got to make a userscript to turn it into desired printable output before I got fed up with how seemingly ordinary tweaks caused strange breakages in the print preview, where the DOM Inspector is very crippled:

```css
/* Stuff that got applied to things I want printed */
noprint, .noprint {
    display: block !important;
}
tabs > tab > panel {
    display: inline !important;
}
explanation { 
    display: inherit !important;
}

/* Hacks to fix 'diagram is shoved to the top-left corner of page 1' bugs 
   revealed by turning off noprint. */

#custom-types + p + datum { position: relative !important; }
#custom-types + p + datum > .hidden { position: relative !important; }

datum[style='margin-right:70px'] { position: relative !important; }
datum[style='margin-right:70px'] code { white-space: nowrap !important; }

datum.spaced[style='padding-bottom:165px'] memory-entry { 
    position: relative !important; 
}
datum.spaced[style='padding-bottom:165px'] visual { 
    margin-right: 10em;
}

/* Stuff that shouldn't be printed but is (or would be after my tweaks) */
body > div:first-child > a.noprint:first-child { 
    /* Hide the GitHub ribbon that isn't hiding properly */
    display: none !important;
}
page-controls, #helloctrl, subtext::after, tabs.lifetimes + footnotes {
    display: none !important;
}
```

Overall, with the bits like the diagrams and the tab widget unrolling, it feels like I'm reverse-engineering your attempts to fight the CSS layout model instead of working with it.

**Comment by ralfbiedert** [comment]
Hi, thanks for the feedback. Some quick comments:

> The noprint class on the "Fork Me on GitHub ribbon" has no effect.

Yes, this is a bug. It was hidden before but must have come back some commit ago. 

> Entire sections of the reference which I'd want to print are hidden inside <noprint> elements.

This is part issue, part design. Which sections are `noprint` is a bit arbitrary at the moment, happy to discuss. That many of them are `noprint` is because, as you notice, they're not really useful printed. 


> Tab widgets which do print are not unrolled, denying access to all but their default tabs.

This is also part "by design", part issue. The tab solution is nice in that it doesn't require JS to function. On the other hand, I haven't looked into _unrolling_ content. That might not be trivial. If done it might be worth doing that for the online version as well (e.g., have option next to "Ligatures"). 

> Expandable explanations in the memory diagrams print collapsed.

Similar reason.

> If I turn off noprint to access the stuff I want, bits of your diagrams show up in the top-left of page 1 in print mode. 

There is some advanced CSS magic at work there to get these boxes work without JavaScript (it's basically radio buttons on steroids). I don't remember the CSS too well, but it might be possible to have some JS that changes their `display` as boxes. 


> suggest using inline SVG for those rather than 

Yes, I looked into that a bit, but concluded HTML + CSS is nicer to maintain eventually.


> Overall, with the bits like the diagrams and the tab widget unrolling, it feels like I'm reverse-engineering your attempts to fight the CSS layout model instead of working with it.

I assume what you want in the end is the whole page with all boxes expanded and vertically stacked?

**Comment by ralfbiedert** [comment]
Added a button to expand everything, it's below the top bar. 

- Can you test that it works for you?
- Also, I haven't touched the `noprint` areas yet, which ones would you like to include in a PDF?

**Comment by ssokolow** [comment]
> Can you test that it works for you?

It looks good. The one aesthetic tweak I'd suggest would be to remove the `width: 100%` on the tab section headers so it gives a visual cue that the readers are looking at un-stacked "index cards" from a tab widget.

![Screenshot_2020-12-16 Rust Language Cheat Sheet](https://user-images.githubusercontent.com/46915/102353941-b9894e00-3fa1-11eb-8fdc-b38e41ed261d.png)

(Heck, there's a certain aesthetic appeal to that on its own. Some systems, such as fimfiction.net's custom BBcode renderer, use that "tabbed index card" look for cited blockquotes.)

> Also, I haven't touched the `noprint` areas yet, which ones would you like to include in a PDF?

Ideally, I'd like to have a choice, so users who might be in a "laptop with one monitor and no WiFi" situation have the option of printing the whole document and sticking it in a duotang. (My general rule is not to second-guess what users might want to read in print. Give them the choice and follow the principle of least surprise.)

Overall, the biggest problem I have with them is that the use of `noprint` is not intuitive. You see this wall of content and you're likely to not realize something's missing until you've already used up paper and ink and are wondering "Waitaminute. I know I saw _______. What gives?"

On that front, I'd make two suggestions:

1. Default to never hiding actual content that's visible to the user in non-print mode... just stuff that is unarguably unhelpful in print like the ↕️ indicators on the expander titles.
2. Clearly document how to craft custom subsets for print in the Printing & PDF section. (Maybe have every section header contain an on-hover "remove this section" button which affects both screen and print rendering.)

For the pre-rendered version, maybe two PDFs? One with what you think is a good selection and one with everything, using the CSS `page-break-before` property to ensure the sections of limited utility start and end on page boundaries. That way, you can just mention in the instructions that the full PDF is formatted to make printing only the desired subsets easy.

Speaking of ↕️, it feels aesthetically disharmonious in the font my browser is using, so it'd be nice to omit it from prints or use something like ⬍ instead... though, semantically, it's not the right symbol to use in any case. A double-headed arrow like that connotes "grab and drag to resize". You'd at least want ⇵ which implies a two-state behaviour instead of an analog adjustment. Ideally, you'd want to reinforce the consistent use of the iconography users are familiar with from expanders elsewhere, such as Microsoft's ⊞ which turns into ⊟ or Apple's rotating triangle that browsers took inspiration from in their default rendering of `<details>`.

> There is some advanced CSS magic at work there to get these boxes work without JavaScript (it's basically radio buttons on steroids). I don't remember the CSS too well, but it might be possible to have some JS that changes their display as boxes.

I recognized that trick in use for the tab boxes (I'm big on not using JavaScript unnecessarily and I've used it myself) but I don't see why it would be necessary for it to be relevant to how diagrams react to switching from screen to print.

> Yes, I looked into that a bit, but concluded HTML + CSS is nicer to maintain eventually.

I think we'll have to agree to disagree there.

Plus, your diagrams do not look good in Firefox's Accessibility Inspector. I'd suggest using some ARIA attributes to replicate how `<img src="..." alt="diagram" />` looks to a screen reader. (ie. Add some text saying "diagram" that's only visible to screen readers, indicate the purpose of the root node of each diagram, and instruct screen readers to exclude the internal nodes from their reading of the DOM.)

...or, instead of `alt="diagram"`, put `title="synopsis of what the diagram communicates"` on the outermost node.

**Comment by ralfbiedert** [comment]
Ok, I think this issue should be fixed now, closing. The new PDF should be available by now via the page. Also see details below. There was a bit of collateral damage with Firefox not being able to properly print the page anymore, but Chrome still works. PRs for that are welcome, but I'd recommend to just download the auto generated PDF anyway. 

> I'd suggest would be to remove the width: 100%

Fixed, agreed it looks better.


>  I'd like to have a choice

Selecting sections as a first-class feature is probably out of scope unless I get more requests for that (PDF is about 1-2% of all page loads).

If someone really needs that I recommend to remove DOM nodes in the page inspector manually and then reprint (or just clone and edit the markdown). 

>  I have with them is that the use of noprint is not intuitive. [...]"

Fair enough. Most content should be no noprint now. Notable exceptions: TOC, Links & Services sections. 


> Speaking of ↕️, it feels aesthetically disharmonious in the font my browser is using, so it'd be nice to omit it from prints or use something like ⬍ instead...

Not that I disagree in specific cases, but that's more a browser issue to me (why can't I change the icon set?). I tried other Unicode symbols before, but they don't stand out enough on their own. Fontawesome & co were too heavyweight for my taste. That said, I'm willing to accept PRs if they look "equal or better" and are similarly lightweight as emojis are. Details might need to be discussed. 

> Plus, your diagrams do not look good in Firefox's Accessibility Inspector.

Also accepting PRs for that. 

Feel free to comment / reopen if there are still issues regarding the PDF generation.
